// @wordpress/components/build/z-stack/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ZStack = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _getValidChildren = require("../ui/utils/get-valid-children");

var _context = require("../ui/context");

var _styles = require("./styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function UnconnectedZStack(props, forwardedRef) {
  const {
    children,
    className,
    isLayered = true,
    isReversed = false,
    offset = 0,
    ...otherProps
  } = (0, _context.useContextSystem)(props, 'ZStack');
  const validChildren = (0, _getValidChildren.getValidChildren)(children);
  const childrenLastIndex = validChildren.length - 1;
  const clonedChildren = validChildren.map((child, index) => {
    const zIndex = isReversed ? childrenLastIndex - index : index;
    const offsetAmount = offset * index;
    const key = (0, _element.isValidElement)(child) ? child.key : index;
    return (0, _element.createElement)(_styles.ZStackChildView, {
      isLayered: isLayered,
      offsetAmount: offsetAmount,
      zIndex: zIndex,
      key: key
    }, child);
  });
  return (0, _element.createElement)(_styles.ZStackView, (0, _extends2.default)({}, otherProps, {
    className: className,
    ref: forwardedRef
  }), clonedChildren);
}
/**
 * `ZStack` allows you to stack things along the Z-axis.
 *
 * ```jsx
 * import { __experimentalZStack as ZStack } from '@wordpress/components';
 *
 * function Example() {
 *   return (
 *     <ZStack offset={ 20 } isLayered>
 *       <ExampleImage />
 *       <ExampleImage />
 *       <ExampleImage />
 *     </ZStack>
 *   );
 * }
 * ```
 */


const ZStack = (0, _context.contextConnect)(UnconnectedZStack, 'ZStack');
exports.ZStack = ZStack;
var _default = ZStack;
exports.default = _default;
